source/distributed/fully_distributed_tria.cc

```diff
@@ -301,8 +313,140 @@ namespace parallel
+                 active_cells.end())
+          {
+            // nothing has happened with this cell: nothing to do (?)
+            cell->set_subdomain_id(cell->level_subdomain_id());
+          }
+        else
+          {
+            // cell has become active, since it children have been coarsened
+            // make level_subdomain_id to subdomain_id
+            cell->set_subdomain_id(cell->level_subdomain_id());
+          }
+
+      // 3b) update artificial cells (keep only one ghost layer)
+      {
+        // collect vertices belonging to active local cells
+        auto add_vertices_of_cell_to_vertices_owned_by_loclly_owned_cells =
+          [](TriaIterator<CellAccessor<dim, spacedim>> &cell,
+             std::vector<bool> &vertices_owned_by_loclly_owned_cells) {
+            // add vertices belonging to a periodic neighbor
+            for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell; i++)
+              if (cell->has_periodic_neighbor(i))
+                {
+                  const auto face_t = cell->face(i);
+                  const auto face_n = cell->periodic_neighbor(i)->face(
+                    cell->periodic_neighbor_face_no(i));
+                  for (unsigned int j = 0;
+                       j < GeometryInfo<dim>::vertices_per_face;
+                       j++)
+                    {
+                      vertices_owned_by_loclly_owned_cells[face_t->vertex_index(
+                        j)] = true;
+                      vertices_owned_by_loclly_owned_cells[face_n->vertex_index(
+                        j)] = true;
+                    }
+                }
+
+            // add local vertices
+            for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                 v++)
+              vertices_owned_by_loclly_owned_cells[cell->vertex_index(v)] =
+                true;
+          };
+
+        std::vector<bool> vertices_owned_by_loclly_owned_cells(
+          this->n_vertices());
+        for (auto cell : this->cell_iterators())
+          if (cell->active() &&
+              (cell->is_locally_owned() ||
+               cell->level_subdomain_id() == this->locally_owned_subdomain()))
+            add_vertices_of_cell_to_vertices_owned_by_loclly_owned_cells(
+              cell, vertices_owned_by_loclly_owned_cells);
+
+        auto is_locally_relevant = [&](auto &cell) {
+          for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+               v++)
+            if (vertices_owned_by_loclly_owned_cells[cell->vertex_index(v)])
+              return true;
+          return false;
+        };
+
+        // mark all active cells which are not connected via a vertex to
+        // an active local cells as artificial
+        for (auto cell : this->cell_iterators())
+          if (cell->active() && !is_locally_relevant(cell))
+            cell->set_subdomain_id(numbers::artificial_subdomain_id);
```

Comment:
you probably want to mark active cells which are artificial for coarsening and try to coarsen them?